### ISSUE TITLE
Make opaque `Scalar` conversions explicit rather than using `From`/`Into` trait.

### ISSUE THREAD

**Issue by JayWhite2357** [spaceandtimefdn/sxt-proof-of-sql#228]
# Background and Motivation
Currently, the `Scalar` trait requires various conversions using the `From` and `Into` traits. Some of these are confusing, while others are clear.
In general, the rule should be that for types with a natural embedding into the associated `Scalar` field, `Scalar` should implement `From`. In all other cases, the conversion should either be auto-implemented, or a trait method, where appropriate.

**NOTE**: The following issue also necessitates changes to Scalar, so be aware of potential merge conflicts:
https://github.com/spaceandtimelabs/sxt-proof-of-sql/issues/234

# Changes Required

Each of these should be a separate PR.
- [ ] Remove `Into<[u64; 4]>`, `From<[u64; 4]>`, and `RefInto<[u64; 4]>` bounds.
    - Replace these with the trait methods `fn from_limbs(val: [u64; 4]) -> Self` and `fn to_limbs(&self) -> [u64; 4]`
    - These three conversions are the base conversions that are used for many other conversions. A `Scalar` is a field element (number mod some prime) with slightly less than 256 bits. Usually, `Scalar`s are internally stored in Montgomery form, so these conversions are non-free. However, they are needed in some situations.
- [ ] Remove `From<&'a String>`, `From<String>`, and `From<&'a str>` bounds. These three implementations ultimately rely on the conversions to and from limbs. As a result, these should have blanket/default implementations.
    - Replace these with `fn from_str_via_hash(val: &str) -> Self` and add a default implementation.
    - While it would be nice to `impl<S: Scalar> From<String> for S`, this is not allowed by the Rust type system. Solutions other than the default implementation of a trait method are welcome.
    - This conversion requires that the string is hashed in order to convert it to 256 bits which are ultimately converted into the `Scalar` itself. This is different from a parsing method.
- [ ] Remove the `VarInt` bound and replace it with a blanket `impl<S: Scalar> VarInt for S` implementation. The existing (and requested) implementation relies on the limb conversions only.

**Comment by maujim** [comment]
@JayWhite2357 can you please take a look at the liked PR?

**Comment by Dustin-Ray** [comment]
> [@JayWhite2357](https://github.com/JayWhite2357) can you please take a look at the liked PR?

@maujim Sorry for the delay on this, the PR is still open and needs conflict resolution, I can review ASAP if you still want to try to get it in. If you still want to try this I think its possible to resolve the formatter error. This issue might need to be resolved in order to accompany #229 

Let us know when you can

**Comment by maujim** [comment]
@Dustin-Ray I'm on macos / arm and I'm getting two compilation errors on `main` related to my machine. 
Given this, I don't think I'll be diving back into this one. 

```
error: failed to run custom build command for `blitzar-sys v1.99.0`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_DEV_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/Users/mukund/dev/sxt-proof-of-sql/target/debug/build/blitzar-sys-a863c08174abcbdf/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at /Users/mukund/.config/cargo/registry/src/index.crates.io-6f17d22bba15001f/blitzar-sys-1.99.0/build.rs:14:9:
  Unsupported OS. Only Linux is supported.
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/panicking.rs:665:5
     1: core::panicking::panic_fmt
               at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/panicking.rs:76:14
     2: build_script_build::main
     3: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish...
error: failed to run custom build command for `halo2curves v0.8.0`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_DEV_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/Users/mukund/dev/sxt-proof-of-sql/target/debug/build/halo2curves-0db72ce93f44cb27/build-script-build` (exit status: 1)
  --- stderr
  Currently feature `asm` can only be enabled on x86_64 arch.
```

**Comment by Dustin-Ray** [comment]
@maujim _I think_ You can fix this error by turning off blitzar using this env variable:

```bash
export BLITZAR_BACKEND=cpu
```

**Comment by maujim** [comment]
I tried a fresh `git clone && cargo build` and I get the halo2curves issue before the blitzar one.

Took a quick look and it seems that the `arm` feature is enabled by default via the nova-snark dependency. 
https://github.com/microsoft/Nova/blob/main/Cargo.toml#L70

So I think its best for someone else to take over for now

**Comment by blairmichaelg** [comment]
/attempt #228

Hey team, I'm taking this on. Here is my implementation plan:

1. **Centralize Explicit Constructors:** I will add explicit associated functions (e.g., \from_u8, rom_i64, rom_limbs_u64) directly to the Scalar / ScalarExt traits and implement them for the concrete scalar backends, mirroring the current From semantics exactly. 2. Refactor Call Sites: I will systematically replace all implicit .into() calls across the codebase, starting with the highest-fanout hubs like LiteralValue::to_scalar, and then moving through the proof primitives and SQL engine. 3. Remove Implicit Traits: Once all call sites are using explicit constructors and the tests pass, I will remove the generic impl From for Scalar blocks to enforce explicit conversions going forward.

**Time Estimate:** I expect to have a PR ready for review within 48 hours.

**Comment by blairmichaelg** [comment]
Quick update for the team: I'm still actively working on this. Ripping out the implicit \From\ traits exposed a massive web of lazy \.into()\ calls across the test suites and proof primitives. I'm currently running an automated AST/Regex sweep to systematically patch the remaining ~1,000 type-inference errors to ensure the explicit constructors are properly mapped without breaking the math logic. Will have a clean PR up shortly.

**Comment by selenaalpha77-sketch** [comment]
/attempt #228

Part 3 of 3: PR #1168 — blanket `impl<S: Scalar> VarInt for S`

**Comment by AlexMelanFromRingo** [comment]
/attempt #228

Picking up the first checkbox from #1117 (closed by stale-bot, @shruti2522 unable to continue per the thread — @Dustin-Ray invited a pickup). Original refactor commit is preserved with @shruti2522 as author; my follow-up commit handles the Rust 1.91 clippy regressions (`clippy::allow_attributes` on `RefInto`, `clippy::semicolon_if_nothing_returned` on the new `from_limbs` call sites in `bench_append_rows.rs`).

PR is up at #1169. Locally green: `source scripts/run_ci_checks.sh`, `cargo clippy --all-targets --all-features -- -D warnings`, `cargo fmt --check`, and `cargo test --workspace --lib` (1419 + 107 = 1526 lib tests passing).

Leaving the other two checkboxes (`from_str_via_hash` default impl, blanket `VarInt` impl) for separate PRs per the issue's guidance.

**Comment by AlexMelanFromRingo** [comment]
Followup: also picking up the **second** checkbox (`from_str_via_hash`) from #1144 (also closed stale, @javery556 inactive) — PR #1170 is up. Stacked on #1169 so the diff is clean if both land in order. The third checkbox (blanket VarInt impl) is intentionally left for a separate PR per the issue's guidance.

**Comment by qiushuang0115-beep** [comment]
Hi, I’d like to work on the first PR: replacing the limb conversion bounds with explicit from_limbs/to_limbs methods.

**Comment by goelo** [comment]
Opened #1354 for the first limb-conversion part of this issue: removing the Scalar trait bounds on From/Into/RefInto<[u64; 4]> and replacing generic usage with explicit Scalar::from_limbs / Scalar::to_limbs methods.

**Comment by goelo** [comment]
Correction to my previous comment: opened #1397 for the third checkbox, removing VarInt as a Scalar super-bound and replacing the concrete MontScalar varint implementation with a blanket impl for every S: Scalar. It is stacked on #1354 for the explicit limb conversion methods and can be rebased once a limb-conversion PR lands.

**Comment by LMSlay** [comment]
/attempt #228

I plan to submit a narrowly scoped first PR for the limb conversion part of this issue: add explicit Scalar::from_limbs and Scalar::to_limbs methods, update MontScalar and call sites that currently rely on From/Into/RefInto<[u64; 4]>, and include focused tests to preserve existing round-trip behavior. I will keep the string conversion and VarInt parts out of this PR as requested.

**Comment by zhaog100** [comment]
👋 I'd like to work on this bounty. Will start with detailed analysis and implementation plan.

**Comment by dashan12344423** [comment]
I’d like to work on the first subtask: replacing the limb conversion bounds with explicit Scalar trait methods. I’ll
  open a focused PR for that part only.

**Comment by ncthuc2004** [comment]
Hi maintainers, I can take this issue and submit a clean PR quickly.

- Stack fit: React/Next.js/TypeScript/Java automation
- ETA: 2-8 hours after confirmation
- Delivery: code + tests + validation notes

If this bounty is open for external contributors, please confirm payout via https://paypal.me/ThucNguyen2004.

**Comment by amithrh** [comment]
/attempt #228

I’m taking a narrowly scoped pass at the first checkbox only: replacing the limb conversion bounds with explicit Scalar trait methods.

Plan:
- Add `Scalar::from_limbs` and `Scalar::to_limbs`.
- Remove the `From<[u64; 4]>`, `Into<[u64; 4]>`, and `RefInto<[u64; 4]>` super-bounds from `Scalar`.
- Update the generic call sites that currently rely on those implicit limb conversions.
- Remove any dead helper code left behind by the refactor.

I’ll leave `from_str_via_hash` and the blanket `VarInt` impl for separate PRs, as requested in the issue. ETA: PR today after final check/fmt validation.

**Comment by liangtovi-debug** [comment]
/attempt #228

Taking a narrowly scoped first-slice PR for the limb conversion checkbox only:
1. Remove `Into<[u64; 4]>`, `From<[u64; 4]>`, and `RefInto<[u64; 4]>` bounds from `Scalar`.
2. Add explicit `from_limbs`/`to_limbs` trait methods and wire implementations for existing scalar types.
3. Update call sites in this slice to use explicit methods, then run focused no-default proof-of-sql tests.

**Comment by liangtovi-debug** [comment]
/claim #228

Submitted focused first-slice PR: https://github.com/spaceandtimefdn/sxt-proof-of-sql/pull/1940

Scope completed:
- Added explicit `Scalar::from_limbs` / `Scalar::to_limbs` methods
- Removed `Into<[u64; 4]>` and `From<[u64; 4]>` as `Scalar` trait bounds
- Updated internal limb-conversion call sites in this slice to use explicit methods

Validation run:
- `cargo test -p proof-of-sql --no-default-features --features test scalar_ext -- --nocapture`
- `cargo test -p proof-of-sql --no-default-features --features test literal_value -- --nocapture`

**Comment by NAEL592** [comment]
/attempt #228

I will audit the opaque Scalar From/Into conversion sites, replace them with explicit conversion calls in a source-compatible way where possible, and run the focused Rust test/check suite.

**Comment by MyTH-zyxeon** [comment]
Maintainer-facing acceptance synthesis for #228. I am not claiming this bounty and did not open a competing PR.

Current shape of the issue:

- #228 explicitly asks for separate PRs for the three slices: limb conversions, string hashing, and blanket VarInt.
- The issue is still open and there are many overlapping historical attempts, so the useful next step is not another broad implementation; it is selecting one clean PR per requested slice.
- Visible current candidates are mergeable but still `mergeStateStatus=BLOCKED` / `reviewDecision=REVIEW_REQUIRED`, with only Orca checks visible from the contributor side.

Suggested landing map:

- First checkbox, limb conversions: #1940 is a focused current candidate. It adds `Scalar::from_limbs` / `Scalar::to_limbs`, removes the generic limb conversion super-bounds including the later `RefInto<[u64; 4]>` gap, and keeps concrete compatibility impls on `MontScalar` while moving generic call sites to explicit methods. Older #1169 covers a similar lane with broader validation, so maintainers should decide whether #1940 is the preferred smaller current landing PR or whether #1169 remains the canonical first-slice attempt.
- Second checkbox, string hashing: #2091 is the visible focused candidate. It removes generic string conversion super-bounds, adds `Scalar::from_str_via_hash`, and updates generic varchar/literal conversion paths. It notes that full cargo/rustfmt validation still needs maintainer CI.
- Third checkbox, blanket VarInt: #2189 is the visible current candidate. It removes `VarInt` as a `Scalar` supertrait bound, adds `impl<S: Scalar> VarInt for S`, and includes focused varint/zigzag validation commands.

Acceptance checks I would use before rewarding or merging any slice:

- Confirm each selected PR only closes its own checkbox and does not accidentally claim all of #228.
- Run maintainer CI or request `run-ci` on the chosen PRs; contributor-visible checks only show the Orca scan set.
- For the limb slice, confirm no generic `Scalar` bound still requires `Into<[u64; 4]>`, `From<[u64; 4]>`, or `RefInto<[u64; 4]>`.
- For the string slice, confirm `from_str_via_hash` preserves the existing byte-hash semantics and is not confused with parsing.
- For the VarInt slice, confirm the blanket implementation preserves the existing U256 representation and zigzag round trips.
- Decide whether concrete compatibility impls are intentionally retained while generic APIs stop depending on them.
- Avoid duplicate reward confusion by selecting one PR per checkbox and leaving the other overlapping attempts unmerged or explicitly superseded.

This is only a triage/landing synthesis to help converge the crowded #228 queue.

**Comment by heickerv1001-dev** [comment]
Prepared PR for #228: https://github.com/spaceandtimefdn/sxt-proof-of-sql/pull/2243

Scope covered: explicit `Scalar::from_limbs` / `Scalar::to_limbs` API, removal of opaque limb conversion bounds, and the associated call-site updates.

**Comment by heickerv1001-dev** [comment]
Published the second #228 slice as a separate PR: explicit `Scalar::from_str_via_hash` plus varchar call-site updates only. The limb-conversion slice remains in PR #2243, so this PR stays scoped to the string-hash checkbox.

**Comment by tonghuaxingdsb** [comment]
/attempt #228

I am taking the remaining non-overlapping VarInt slice from the checklist:

- remove the `VarInt` supertrait bound from `Scalar`
- replace the `MontScalar<T>`-specific `VarInt` implementation with a blanket `impl<S: Scalar> VarInt for S`
- keep the existing scalar varint behavior by routing the helper functions through the existing limb-based conversions

This intentionally does not touch the limb-conversion or string-hash slices already covered by other PRs.

**Comment by tonghuaxingdsb** [comment]
Submitted the non-overlapping VarInt slice here: https://github.com/spaceandtimefdn/sxt-proof-of-sql/pull/2439

**Comment by chiddekel** [comment]
/attempt #228

Taking the first checkbox only — the limb-conversion slice:
- Add explicit `Scalar::from_limbs(val: [u64; 4]) -> Self` and `Scalar::to_limbs(&self) -> [u64; 4]` trait methods.
- Remove the `Into<[u64; 4]>`, `From<[u64; 4]>`, and `RefInto<[u64; 4]>` super-bounds from `Scalar`.
- Migrate the generic `S: Scalar` call sites to the explicit methods, keeping the concrete `MontScalar` impls for compatibility.

Leaving the `from_str_via_hash` and blanket `VarInt` checkboxes for separate PRs as the issue requests.

**Comment by JayWhite2357** [comment]
Closing all issues due to suspected low-effort AI slop/spam.
